### PR TITLE
feat: [IDP-852]: Update k8s secret only if the last updated timestamp has changed

### PR DESCRIPTION
**Pull request by vikyathharekal**
## Describe your changes
1. Update k8s secret if entity.getSecretLastModifiedAt() < lastModifiedAt for type secret.
2. Update k8s secret if entity.getValue().equals(configEnvVariable.getValue() for type config.
3. Add migration to add secretLastModifiedAt field.
4. Other changes 
     Fixed - The token used for authorization could be missing. There is no way to figure this out in logs directly now.
     Fixed - Failed to retrieve decrypted value after multiple retries error message doesn’t log the secret identifier. We have to look at the info logs to figure it out.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53727)

### CODE REVIEW

File: idp-service/src/main/java/io/harness/idp/migration/BackstageEnvSecretSecretLastModifiedAtMigration.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    BasicDBObject updateOps = new BasicDBObject(BackstageEnvSecretVariableKeys.secretLastModifiedAt, -1);
+    BulkWriteOperation writeOperation =
+        mongoPersistence.getCollection(BackstageEnvVariableEntity.class).initializeUnorderedBulkOperation();
+    writeOperation.find(basicDBObject).update(new BasicDBObject("$set", updateOps));
+    BulkWriteResult updateOperationResult = writeOperation.execute();
+    if (updateOperationResult.getModifiedCount() > 0) {
+      log.info(
+          "Added secretLastModifiedAt field successfully for {} records", updateOperationResult.getModifiedCount());
+    } else {
+      log.warn("Could not add isDeleted field to any record");
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/java/io/harness/idp/migration/BackstageEnvSecretSecretLastModifiedAtMigration.java:44]: nit: Could not add secretLastModifiedAt field with default value as -1 to any record

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by vikyathharekal** [comment]
trigger uts

**Review by sathish-soundarapandian — APPROVED** [review]

**Review by devesh-harness — APPROVED** [review]

**Comment by vikyathharekal** [comment]
trigger review

**Comment by vikyathharekal** [comment]
trigger ut7

File: idp-service/src/main/java/io/harness/idp/migration/BackstageEnvSecretSecretLastModifiedAtMigration.java (hunk 1 above)
Comment (vikyathharekal) [reply idp-service/src/main/java/io/harness/idp/migration/BackstageEnvSecretSecretLastModifiedAtMigration.java]: Fixed

**Pull request merged**